pkg/bloomcompactor/sharding.go

```diff
@@ -0,0 +1,73 @@
+package bloomcompactor
+
+import (
+	"github.com/grafana/dskit/ring"
+)
+
+var (
+	// TODO: Should we include LEAVING instances in the replication set?
+	RingOp = ring.NewOp([]ring.InstanceState{ring.JOINING, ring.ACTIVE}, nil)
+)
+
+// ShardingStrategy describes whether compactor "owns" given user or job.
+type ShardingStrategy interface {
+	OwnsTenant(tenant string) (bool, error)
+	OwnsJob(job Job) (bool, error)
+}
+
+type ShuffleShardingStrategy struct {
+	ring           *ring.Ring
+	ringLifeCycler *ring.BasicLifecycler
+	limits         Limits
+
+	// Buffers to avoid allocations in ring.Get
+	bufDescs           []ring.InstanceDesc
+	bufHosts, bufZones []string
+}
+
+func NewShuffleShardingStrategy(r *ring.Ring, ringLifecycler *ring.BasicLifecycler, limits Limits) *ShuffleShardingStrategy {
+	s := ShuffleShardingStrategy{
+		ring:           r,
+		ringLifeCycler: ringLifecycler,
+		limits:         limits,
+	}
+	s.bufDescs, s.bufHosts, s.bufZones = ring.MakeBuffersForGet()
```

Comment:
Usage of these buffers is not threadsafe

Reply:
Indeed. Thank you! If we end up making many allocations here, we can use a buffer pool.